# Comprehensions

Las comprensiones de Python son métodos concisos para crear nuevas secuencias (listas, conjuntos, diccionarios) mediante operaciones con secuencias existentes.

Permiten escribir código compacto y legible para generar secuencias basadas en ciertas condiciones o transformaciones.

Existen tres tipos de comprensiones en Python: de lista, de conjunto y de diccionario.


## List comprehensions
Las comprensiones de listas ofrecen una forma concisa de crear nuevas listas a partir de listas existentes u otros objetos iterables.

Syntax:

new_list = [expression for item in iterable if condition]

Donde:
- expression: La operación o transformación que se aplicará a cada elemento.
- item: El elemento del iterable que se está procesando.
- iterable: La lista, cadena o cualquier otro objeto iterable existente.
- condition (opcional): Una condición que filtra los elementos según un criterio determinado.

Veamos algunos ejemplos.

In [1]:
base_list = [1, 2, 3, 4, 5, 6, 7, 8]

In [2]:
[x for x in base_list]

[1, 2, 3, 4, 5, 6, 7, 8]

Apliquemos algunas transformaciones.

In [ ]:
[x ** 2 for x in base_list]

In [ ]:
[x > 5 for x in base_list]

Podemos crear tipos estructurados dentro de la lista resultante

In [3]:
[(x, x**2) for x in base_list]

[(1, 1), (2, 4), (3, 9), (4, 16), (5, 25), (6, 36), (7, 49), (8, 64)]

In [4]:
[[x, x+1, x+2] for x in base_list]

[[1, 2, 3],
 [2, 3, 4],
 [3, 4, 5],
 [4, 5, 6],
 [5, 6, 7],
 [6, 7, 8],
 [7, 8, 9],
 [8, 9, 10]]

In [5]:
["<>" * x for x in base_list]

['<>',
 '<><>',
 '<><><>',
 '<><><><>',
 '<><><><><>',
 '<><><><><><>',
 '<><><><><><><>',
 '<><><><><><><><>']

Filtrar los resultados

In [6]:
base_list

[1, 2, 3, 4, 5, 6, 7, 8]

In [7]:
[x for x in base_list if x > 4]

[5, 6, 7, 8]

In [8]:
[x for x in base_list if x % 2 == 0]

[2, 4, 6, 8]

### Nested comprehensions
También puedes anidar comprensiones, donde cada elemento que creas utiliza otra comprensión.

In [9]:
[[v for v in range(x)] for x in base_list]

[[0],
 [0, 1],
 [0, 1, 2],
 [0, 1, 2, 3],
 [0, 1, 2, 3, 4],
 [0, 1, 2, 3, 4, 5],
 [0, 1, 2, 3, 4, 5, 6],
 [0, 1, 2, 3, 4, 5, 6, 7]]

In [10]:
[[row * col for col in range(1, 4)] for row in range(1, 4)]

[[1, 2, 3], [2, 4, 6], [3, 6, 9]]

## Chained comprehensions

In [11]:
# Flattening a nested list
nested_list = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]

[v for sublist in nested_list for v in sublist]

[1, 2, 3, 4, 5, 6, 7, 8, 9]

In [12]:
# Creating a list of tuples
numbers = [1, 2, 3, 4, 5]
pairs = [(x, y) for x in numbers for y in numbers if x != y]
print(pairs)

[(1, 2), (1, 3), (1, 4), (1, 5), (2, 1), (2, 3), (2, 4), (2, 5), (3, 1), (3, 2), (3, 4), (3, 5), (4, 1), (4, 2), (4, 3), (4, 5), (5, 1), (5, 2), (5, 3), (5, 4)]


## Dictionary Comprehensions
Las comprensiones de diccionario en Python permiten crear diccionarios de forma concisa y expresiva. Siguen una sintaxis similar a la de las comprensiones de lista, pero generan diccionarios como resultado.

Syntax:

new_dict = {key_expression: value_expression for item in iterable if condition}

In [13]:
values = [x for x in range(10)]
values

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

In [14]:
{x: 0 for x in values}

{0: 0, 1: 0, 2: 0, 3: 0, 4: 0, 5: 0, 6: 0, 7: 0, 8: 0, 9: 0}

In [15]:
{x: x**2 for x in values}

{0: 0, 1: 1, 2: 4, 3: 9, 4: 16, 5: 25, 6: 36, 7: 49, 8: 64, 9: 81}

In [16]:
{(x, x % 2 == 0): [x, x+1] for x in values}

{(0, True): [0, 1],
 (1, False): [1, 2],
 (2, True): [2, 3],
 (3, False): [3, 4],
 (4, True): [4, 5],
 (5, False): [5, 6],
 (6, True): [6, 7],
 (7, False): [7, 8],
 (8, True): [8, 9],
 (9, False): [9, 10]}

In [17]:
stone_names = ["Quartz", "Amethyst", "Diamond", "Ruby", "Emerald", "Sapphire", "Topaz", "Opal", "Jade", "Citrine"]

In [18]:
# Asociar nombre con longitud
{name: len(name) for name in stone_names}

{'Quartz': 6,
 'Amethyst': 8,
 'Diamond': 7,
 'Ruby': 4,
 'Emerald': 7,
 'Sapphire': 8,
 'Topaz': 5,
 'Opal': 4,
 'Jade': 4,
 'Citrine': 7}

In [19]:
# contar vocales
{name: sum(1 for letter in name.lower() if letter in 'aeiou') for name in stone_names}

{'Quartz': 2,
 'Amethyst': 2,
 'Diamond': 3,
 'Ruby': 1,
 'Emerald': 3,
 'Sapphire': 3,
 'Topaz': 2,
 'Opal': 2,
 'Jade': 2,
 'Citrine': 3}

In [20]:
# ¿Contiene 'a'?
{name: 'Yes' if 'a' in name.lower() else 'No' for name in stone_names}

{'Quartz': 'Yes',
 'Amethyst': 'Yes',
 'Diamond': 'Yes',
 'Ruby': 'No',
 'Emerald': 'Yes',
 'Sapphire': 'Yes',
 'Topaz': 'Yes',
 'Opal': 'Yes',
 'Jade': 'Yes',
 'Citrine': 'No'}

In [ ]:
# Reverse names
{name: name[::-1] for name in stone_names}

También se pueden utilizar filtros

In [21]:
{s: len(s) for s in stone_names if s[0] in "AEIOU"}

{'Amethyst': 8, 'Emerald': 7, 'Opal': 4}

In [22]:
{name: len(name) for name in stone_names if len(name) > 5}

{'Quartz': 6,
 'Amethyst': 8,
 'Diamond': 7,
 'Emerald': 7,
 'Sapphire': 8,
 'Citrine': 7}

In [23]:
{name: sum(1 for letter in name.lower() if letter in 'aeiou') for name in stone_names if name[0].lower() in 'aeiou'}

{'Amethyst': 2, 'Emerald': 3, 'Opal': 2}

In [24]:
# Sólo para palíndromos
words = ['opal', 'emerald', 'ruby', 'madam', 'sapphire', 'level', 'diamond']
{name: None for name in words if name.lower() == name.lower()[::-1]}


{'madam': None, 'level': None}

## Ejercicios Resueltos


### List Comprehensions

**Ejercicio**. Crea una lista idéntica a la primera lista usando la comprensión de listas.

In [25]:
l = [2, 3, 4, 5, 6]
[x for x in l]

[2, 3, 4, 5, 6]

**Ejercicio**. Crea una lista con los elementos de un rango de 1200 a 2000, con incrementos de 130, usando comprensión de listas.

In [26]:
[x for x in range(1200, 2001, 130)]

[1200, 1330, 1460, 1590, 1720, 1850, 1980]

**Ejercicio**. Usa la comprensión de listas para construir una nueva lista, pero suma 6 a cada elemento.

In [27]:
l = [2, 3, 4, 5, 6]
[x + 6 for x in l]

[8, 9, 10, 11, 12]

**Ejercicio**. Usando la comprensión de listas, construye una lista con los cuadrados de cada elemento.

In [28]:
l = [2, 4, 6, 8, 10, 12, 14]
[x ** 2 for x in l]

[4, 16, 36, 64, 100, 144, 196]

**Ejercicio**. Usando la comprensión de listas, construye una lista con los cuadrados de cada elemento, si el cuadrado es mayor que 50.


In [29]:
l = [2, 4, 6, 8, 10, 12, 14]
[x ** 2 for x in l if x**2 > 50]

[64, 100, 144, 196]

a) Modificar la solución para evitar calcular la potencia dos veces

In [30]:
[v for v in [x**2 for x in l] if v > 50]

[64, 100, 144, 196]

**Ejercicio**. El diccionario contiene vehículos y sus pesos en kilogramos. Construya una lista con los nombres de vehículos con un peso inferior a 5000 kilogramos. En la misma lista de comprensión, escriba todos los nombres clave en mayúsculas.


In [31]:
dict={"Sedan": 1500, "SUV": 2000, "Pickup": 2500, "Minivan": 1600, 
      "Van": 2400, "Semi": 13600, "Bicycle": 7, "Motorcycle": 110}
[k.upper() for k, v in dict.items() if v < 5000]

['SEDAN', 'SUV', 'PICKUP', 'MINIVAN', 'VAN', 'BICYCLE', 'MOTORCYCLE']

**Ejercicio**. Encuentra todos los números del 1 al 1000 que sean divisibles por 8.

In [32]:
[x for x in range(1, 1001) if x % 8 == 0]

[8,
 16,
 24,
 32,
 40,
 48,
 56,
 64,
 72,
 80,
 88,
 96,
 104,
 112,
 120,
 128,
 136,
 144,
 152,
 160,
 168,
 176,
 184,
 192,
 200,
 208,
 216,
 224,
 232,
 240,
 248,
 256,
 264,
 272,
 280,
 288,
 296,
 304,
 312,
 320,
 328,
 336,
 344,
 352,
 360,
 368,
 376,
 384,
 392,
 400,
 408,
 416,
 424,
 432,
 440,
 448,
 456,
 464,
 472,
 480,
 488,
 496,
 504,
 512,
 520,
 528,
 536,
 544,
 552,
 560,
 568,
 576,
 584,
 592,
 600,
 608,
 616,
 624,
 632,
 640,
 648,
 656,
 664,
 672,
 680,
 688,
 696,
 704,
 712,
 720,
 728,
 736,
 744,
 752,
 760,
 768,
 776,
 784,
 792,
 800,
 808,
 816,
 824,
 832,
 840,
 848,
 856,
 864,
 872,
 880,
 888,
 896,
 904,
 912,
 920,
 928,
 936,
 944,
 952,
 960,
 968,
 976,
 984,
 992,
 1000]

**Ejercicio**. Encuentra todos los números del 1 al 1000 que contengan un 6.


In [33]:
[x for x in range(1, 1001) if '6' in str(x)]

[6,
 16,
 26,
 36,
 46,
 56,
 60,
 61,
 62,
 63,
 64,
 65,
 66,
 67,
 68,
 69,
 76,
 86,
 96,
 106,
 116,
 126,
 136,
 146,
 156,
 160,
 161,
 162,
 163,
 164,
 165,
 166,
 167,
 168,
 169,
 176,
 186,
 196,
 206,
 216,
 226,
 236,
 246,
 256,
 260,
 261,
 262,
 263,
 264,
 265,
 266,
 267,
 268,
 269,
 276,
 286,
 296,
 306,
 316,
 326,
 336,
 346,
 356,
 360,
 361,
 362,
 363,
 364,
 365,
 366,
 367,
 368,
 369,
 376,
 386,
 396,
 406,
 416,
 426,
 436,
 446,
 456,
 460,
 461,
 462,
 463,
 464,
 465,
 466,
 467,
 468,
 469,
 476,
 486,
 496,
 506,
 516,
 526,
 536,
 546,
 556,
 560,
 561,
 562,
 563,
 564,
 565,
 566,
 567,
 568,
 569,
 576,
 586,
 596,
 600,
 601,
 602,
 603,
 604,
 605,
 606,
 607,
 608,
 609,
 610,
 611,
 612,
 613,
 614,
 615,
 616,
 617,
 618,
 619,
 620,
 621,
 622,
 623,
 624,
 625,
 626,
 627,
 628,
 629,
 630,
 631,
 632,
 633,
 634,
 635,
 636,
 637,
 638,
 639,
 640,
 641,
 642,
 643,
 644,
 645,
 646,
 647,
 648,
 649,
 650,
 651,
 652,
 653,
 654,
 655,


In [34]:
# Una solución más elegante

def has_digit(n, digit):
    while n > 0:
        r = n % 10
        if r == digit:
            return True
        n = n // 10
    return False

[x for x in range(1,1001) if has_digit(x, 6)]

[6,
 16,
 26,
 36,
 46,
 56,
 60,
 61,
 62,
 63,
 64,
 65,
 66,
 67,
 68,
 69,
 76,
 86,
 96,
 106,
 116,
 126,
 136,
 146,
 156,
 160,
 161,
 162,
 163,
 164,
 165,
 166,
 167,
 168,
 169,
 176,
 186,
 196,
 206,
 216,
 226,
 236,
 246,
 256,
 260,
 261,
 262,
 263,
 264,
 265,
 266,
 267,
 268,
 269,
 276,
 286,
 296,
 306,
 316,
 326,
 336,
 346,
 356,
 360,
 361,
 362,
 363,
 364,
 365,
 366,
 367,
 368,
 369,
 376,
 386,
 396,
 406,
 416,
 426,
 436,
 446,
 456,
 460,
 461,
 462,
 463,
 464,
 465,
 466,
 467,
 468,
 469,
 476,
 486,
 496,
 506,
 516,
 526,
 536,
 546,
 556,
 560,
 561,
 562,
 563,
 564,
 565,
 566,
 567,
 568,
 569,
 576,
 586,
 596,
 600,
 601,
 602,
 603,
 604,
 605,
 606,
 607,
 608,
 609,
 610,
 611,
 612,
 613,
 614,
 615,
 616,
 617,
 618,
 619,
 620,
 621,
 622,
 623,
 624,
 625,
 626,
 627,
 628,
 629,
 630,
 631,
 632,
 633,
 634,
 635,
 636,
 637,
 638,
 639,
 640,
 641,
 642,
 643,
 644,
 645,
 646,
 647,
 648,
 649,
 650,
 651,
 652,
 653,
 654,
 655,


**Ejercicio**. Encuentra todos los números del 1 al 1000 que tengan exactamente dos apariciones del dígito 4.

In [35]:
def count_digit(n, digit):
    result = 0
    while n > 0:
        r = n % 10
        if r == digit:
            result += 1
        n = n // 10
    return result

print([x for x in range(1001) if count_digit(x, 4) == 2])

[44, 144, 244, 344, 404, 414, 424, 434, 440, 441, 442, 443, 445, 446, 447, 448, 449, 454, 464, 474, 484, 494, 544, 644, 744, 844, 944]


**Ejercicio**. Usa la comprensión anidada para encontrar todos los números del 1 al 200 que no sean divisibles por ningún dígito del 2 al 9.


In [36]:
[n for n in range(1, 200) if all(n%d != 0 for d in range(2,10))]

[1,
 11,
 13,
 17,
 19,
 23,
 29,
 31,
 37,
 41,
 43,
 47,
 53,
 59,
 61,
 67,
 71,
 73,
 79,
 83,
 89,
 97,
 101,
 103,
 107,
 109,
 113,
 121,
 127,
 131,
 137,
 139,
 143,
 149,
 151,
 157,
 163,
 167,
 169,
 173,
 179,
 181,
 187,
 191,
 193,
 197,
 199]

**Ejercicio**. Usa una comprensión anidada para encontrar todos los números del 1 al 200 que tengan exactamente dos divisores no triviales.


In [37]:
[n for n in range(1, 201) if sum([1 for d in range(2,n) if n%d == 0]) == 2]

[6,
 8,
 10,
 14,
 15,
 21,
 22,
 26,
 27,
 33,
 34,
 35,
 38,
 39,
 46,
 51,
 55,
 57,
 58,
 62,
 65,
 69,
 74,
 77,
 82,
 85,
 86,
 87,
 91,
 93,
 94,
 95,
 106,
 111,
 115,
 118,
 119,
 122,
 123,
 125,
 129,
 133,
 134,
 141,
 142,
 143,
 145,
 146,
 155,
 158,
 159,
 161,
 166,
 177,
 178,
 183,
 185,
 187,
 194]

a) Qué necesitas modificar en el código para obtener números primos

In [38]:
[n for n in range(1, 201) if sum([1 for d in range(2,n) if n%d == 0]) == 0]

[1,
 2,
 3,
 5,
 7,
 11,
 13,
 17,
 19,
 23,
 29,
 31,
 37,
 41,
 43,
 47,
 53,
 59,
 61,
 67,
 71,
 73,
 79,
 83,
 89,
 97,
 101,
 103,
 107,
 109,
 113,
 127,
 131,
 137,
 139,
 149,
 151,
 157,
 163,
 167,
 173,
 179,
 181,
 191,
 193,
 197,
 199]

**Ejercicio**. Eliminar todas las vocales de una cadena.

In [39]:
string = "Practice Problems to Drill List Comprehension in Your Head."
"".join([c for c in string if c.lower() not in 'aeiou'])

'Prctc Prblms t Drll Lst Cmprhnsn n Yr Hd.'

**Ejercicio**. Encuentra todas las palabras de una cadena que tengan menos de 5 letras (usa la cadena anterior).

In [40]:
[s for s in string.split() if len(s) < 5]

['to', 'List', 'in', 'Your']

**Ejercicio**. Usa una lista de comprensión anidada para encontrar todos los números del 1 al 10000 que sean divisibles por cualquier dígito excepto 1 (2-9).


In [41]:
[x for x in range(1, 10001) if all([x % d == 0 for d in range(2, 10)])]

[2520, 5040, 7560]

**Ejercicio**. Crea una lista de todas las consonantes de la cadena "A los yaks amarillos les gusta gritar y bostezar, y ayer cantaban mientras comían ñames asquerosos".

In [ ]:
phrase = "Yellow Yaks like yelling and yawning and yesterday they yodled while eating yuky yams"
[c for c in phrase if c.lower() not in "aeiou"]

**Exercise**. Get the index and the value as a tuple for items in the list “hi”, 4, 8.99, ‘apple’, (‘t,b’,’n’). Result would look like (index, value), (index, value)

In [ ]:
l = ['hi', 4, 8.99, 'apple', ('t,b','n')]
[(idx, l[idx]) for idx in range(len(l))]

**Exercise**. Find the common numbers in two lists (without using a tuple or set) list_a = 1, 2, 3, 4, list_b = 2, 3, 4, 5

In [ ]:
list_a = [1, 2, 3, 4]
list_b = [2, 3, 4, 5]
[a for a in list_a if a in list_b]

**Exercise**. Given numbers = range(20), produce a list containing the word ‘even’ if a number in the numbers is even, and the word ‘odd’ if the number is odd. Result would look like ‘odd’,’odd’, ‘even’

In [ ]:
['even' if n % 2 == 0 else 'odd' for n in range(20)]

**Exercise**. Create a single string that contains the second-to-last letter of each word in text, sorted alphabetically and in lowercase. If a word is less than two letters in length, use the single character available.

In [ ]:
text = "Alice was beginning to get very tired of sitting \
by her sister on the bank, and of having nothing to do: \
once or twice she had peeped into the book her sister \
was reading, but it had no pictures or conversations in \
it, 'and what is the use of a book,' thought Alice, \
'without pictures or conversations?"

filtered_text = text.replace("'", "").replace(",", "")
selected_letters = [(l[-2] if len(l) > 1 else l).lower() for l in filtered_text.split(" ")]
"".join(sorted(selected_letters))

**Exercice**. Find the average number of characters per word in text, rounded to the nearest hundredth. This value should exclude special characters, such as quotation marks and semicolons. 

In [ ]:
filtered_text = text.replace("'", "").replace(",", "")
character_count = [len(l) for l in filtered_text.split()]
round(sum(character_count) / len(character_count), 2)

**Exercise**. The following code loads a list of Marvel names downloaded from the internet

In [ ]:
with open('data/Marvel-names.txt', 'r', encoding='iso-8859-1',
                 errors='ignore') as f:
    lines = f.readlines()
lines[:10]

a) Create a list with the character names

In [ ]:
def process_line(l):
    idx1 = l.find('"')
    l = l[idx1+1:]
    idx2 = l.find('"')
    return l[:idx2].strip()

names = [process_line(l) for l in lines]
names[1:10]

b) Some characters have multiple names, separated by '/'. Modify the list so that each element becomes a list of these values. Remove empty values

In [ ]:
all_names = [[v for v in n.split("/") if len(v)>0] for n in names]
all_names = [n for n in all_names if len(n) > 0]
all_names[:10]

c) Find the heroes with the largest number of names

In [ ]:
max_cant = max((len(n) for n in all_names))
print(max_cant)
more_names = [v for v in all_names if len(v) == max_cant]
more_names[:30]

d) Remove those characters where the first name contains less than three characters

In [ ]:
selected = [m for m in all_names if len(m[0]) > 2]
selected

e) Add to each name in the list a '%' character in the begining and end

In [ ]:
[['%'+n+'%' for n in m] for m in selected]

### Dictionary comprehensions

**Exercise**. Create a dictionary from the list with same key:value pairs, such as: {"key": "key"}.

In [ ]:
lst=["NY", "FL", "CA", "VT"]
{k: k for k in lst}

**Exercise**. Create a dictionary where each number in the range from 100 to 160 with steps of 10 is the key and each item divided by 100 is the value.

In [ ]:
{v: v/100 for v in range(100, 161, 10)}

**Exercise**. Create a dictionary from the current dictionary where only the key:value pairs with value above 2000 are taken to the new dictionary.

In [ ]:
dict1={"NFLX":4950,"TREX":2400,"FIZZ":1800, "XPO":1700}
{k: v for k, v in dict1.items() if v > 2000}

**Exercise**. Use a dictionary comprehension to count the length of each word in a sentence 

In [ ]:
string = "Practice Problems to Drill List Comprehension in Your Head"
{s: len(s) for s in string.split()}

**Exercise**. For all the numbers 1–1000, use a nested list/dictionary comprehension to find the highest single digit any of the numbers is divisible by

In [ ]:
{x: max([d for d in range(2, 10) if x%d == 0],default=None) for x in range(1, 1001)}

**Exercise**. Write a Python program to convert two lists into a dictionary in a way that item from list1 is the key and item from list2 is the value

In [ ]:
keys = ['Ten', 'Twenty', 'Thirty']
values = [10, 20, 30]
{keys[idx]: values[idx] for idx in range(min(len(keys), len(values)))}

**Exercise**. Initialize dictionary with default values: create a new dictionary with provided keys an default values.

In [ ]:
employees = ['Kelly', 'Emma']
defaults = {"designation": 'Developer', "salary": 8000}

{k: defaults for k in employees}

**Exercise**. Create a dictionary by extracting the keys from a given dictionary

In [ ]:
sample_dict = {
    "name": "Kelly",
    "age": 25,
    "salary": 8000,
    "city": "New york"}

# Keys to extract
keys = ["name", "salary"]

{k: v for k, v in sample_dict.items() if k in keys}

**Exercise**. Delete a list of keys from a dictionary

In [ ]:
sample_dict = {
    "name": "Kelly",
    "age": 25,
    "salary": 8000,
    "city": "New york"
}

# Keys to remove
keys = ["name", "salary"]

{k: v for k, v in sample_dict.items() if k not in keys}